### PR TITLE
Add farzone support

### PR DESCRIPTION
**Pull request by tbak**

### CODE REVIEW

File: titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/KubeNotificationProcessor.java [hunk 1]
```diff
@@ -264,6 +266,16 @@ private Task attachNodeMetadata(Task task, V1Node node) {
 
         acceptNotNull(node.getMetadata().getName(), nodeName -> agentAttributes.put(TaskAttributes.TASK_ATTRIBUTES_KUBE_NODE_NAME, nodeName));
 
+        List<V1NodeAddress> addresses = node.getStatus().getAddresses();
```
Comment (corindwyer) [inline titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/KubeNotificationProcessor.java]: The first ipv4 ip address should be used as the host ip like in:
https://github.com/Netflix/titus-control-plane/blob/master/titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/KubeApiServerIntegrator.java#L453

Comment (tbak) [reply titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/KubeNotificationProcessor.java]: @corindwyer I have seen your recent PR. We should share your address filter in both places. Once you merge, I will rebase and create a function in `KubeUtil`.

Comment (corindwyer) [reply titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/KubeNotificationProcessor.java]: Sounds good. I already merged.

File: titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/KubeUtil.java [hunk 2]
```diff
@@ -191,4 +197,12 @@ public static boolean isOwnedByKubeScheduler(V1Pod v1Pod) {
         }
         return false;
     }
+
+    public static String getNodeIpV4Address(V1Node node) {
+        return   node.getStatus().getAddresses().stream()
```
Comment (corindwyer) [inline titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/KubeUtil.java]: this looks incorrectly indented

File: titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/batch/BatchDifferenceResolver.java [hunk 3]
```diff
@@ -238,7 +244,7 @@ public BatchDifferenceResolver(
         }
 
         Map<String, String> taskContext = getTaskContext(previousTask, unassignedIpAllocations);
-        if (kubeSchedulerPredicate.test(refJobView.getJob().getJobDescriptor())) {
+        if (KubeUtil.findFarzoneId(kubeConfiguration, refJobView.getJob()).isPresent() || kubeSchedulerPredicate.test(refJobView.getJob().getJobDescriptor())) {
```
Comment (fabiokung) [inline titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/batch/BatchDifferenceResolver.java]: why not centralize this check into the `kubeSchedulerPredicate`?

File: titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/service/ServiceDifferenceResolver.java [hunk 4]
```diff
@@ -261,7 +267,7 @@ public ServiceDifferenceResolver(
         }
 
         Map<String, String> taskContext = getTaskContext(previousTask, unassignedIpAllocations);
-        if (kubeSchedulerPredicate.test(refJobView.getJob().getJobDescriptor())) {
+        if (KubeUtil.findFarzoneId(kubeConfiguration, refJobView.getJob()).isPresent() || kubeSchedulerPredicate.test(refJobView.getJob().getJobDescriptor())) {
```
Comment (fabiokung) [inline titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/service/ServiceDifferenceResolver.java]: same as above, centralize this into the predicate

File: titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/KubeUtil.java [hunk 5]
```diff
@@ -148,4 +161,48 @@ public static String formatV1ContainerState(V1ContainerState containerState) {
 
         return "{state=<not set>}";
     }
+
+    /**
+     * If a job has an availability zone hard constraint with a farzone id, return this farzone id.
+     */
+    public static Optional<String> findFarzoneId(DirectKubeConfiguration configuration, Job job) {
+        List<String> farzones = configuration.getFarzones();
```
Comment (fabiokung) [inline titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/KubeUtil.java]: could this be generalized to any availabilityZones? Does it matter that it is a farzone? I'd recommend naming this `isAvailabilityZoneOwnedByKubeScheduler()` and `configuration.getAvailabilityZonesOwnedByKubeScheduler()`, or similar.

IOW, can we abstract away that some AZs are farzones?

File: titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/DefaultPodAffinityFactory.java [hunk 6]
```diff
@@ -181,6 +186,12 @@ private void addNodeAffinityPreferredSelectorConstraint(String key, String value
             term.addMatchExpressionsItem(requirement);
         }
 
+        private void processFarzoneConstraints() {
```
Comment (fabiokung) [inline titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/DefaultPodAffinityFactory.java]: same, generalize to any AZs, farzone or not

File: titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/batch/BatchDifferenceResolver.java (hunk 3 above)
Comment (tbak) [reply titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/batch/BatchDifferenceResolver.java]: I have tried that, but the code is in a different subproject.

File: titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/service/ServiceDifferenceResolver.java (hunk 4 above)
Comment (tbak) [reply titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/service/ServiceDifferenceResolver.java]: I have tried that, but the code is in a different subproject.

File: titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/KubeUtil.java (hunk 5 above)
Comment (tbak) [reply titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/KubeUtil.java]: It is more than an additional zone. Farzones are excluded from the default scheduling, and there is extra logic around that. I would prefer to keep it explicit.

File: titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/DefaultPodAffinityFactory.java (hunk 6 above)
Comment (tbak) [reply titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/DefaultPodAffinityFactory.java]: See my comment above ^^^.

**Pull request merged**